presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -351,34 +370,67 @@ private void invalidateSchemasCache()
         schemaNamesCache.invalidateAll();
     }
 
-    private void invalidateTablesCaches()
+    private void invalidateTablesCaches(SchemaTableName table)
     {
-        columnsCache.invalidateAll();
-        tableHandleCache.invalidateAll();
-        tableNamesCache.invalidateAll();
+        invalidateColumnsCache(table);
+        invalidateTableCache(table);
     }
 
-    private void invalidateColumnsCache(ConnectorSession session, SchemaTableName table)
+    private void invalidateTableCache(SchemaTableName table)
     {
-        columnsCache.invalidate(new ColumnsCacheKey(JdbcIdentity.from(session), table));
+        invalidateCache(tableNamesCache, key -> key.schemaName.equals(Optional.of(table.getSchemaName())));
+        invalidateCache(tableHandleCache, key -> key.tableName.equals(table));
+    }
+
+    private void invalidateColumnsCache(SchemaTableName table)
+    {
+        invalidateCache(columnsCache, key -> key.table.equals(table));
+    }
+
+    private Map<String, Object> getSessionProperties(ConnectorSession session)
+    {
+        return sessionPropertiesProviders.stream()
+                .flatMap(property -> getSessionProperty(session, property))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
```

Comment:
Can session property have a null value?

Reply:
it turns out that yes - in Oracle connector. I've fixed that.

Reply:
@wendigo if null is possible then toImmutableMap will fail. Will you fix it?

Reply:
I've added a singleton object that marks that value is null